### ISSUE TITLE
New Theme: Script-Editor

### ISSUE THREAD

**Issue by Tekl** [HappenApps/Quiver#75]
I would like to see a new theme which looks similar to the Script-Editor.

**Comment by Tekl** [comment]
Okay, I digged into the themes and I tried to change the Xcode theme to the colors for Script Editor. I can't get it look the same as Words like "display dialog" and "return" have the same class ace_type. "return" should be "ace_keyword". Also I can't differentiate between "length" and "character"). :-/

``` javascript
ace.define("ace/theme/xcode",["require","exports","module","ace/lib/dom"],function(e,t,n){t.isDark=!1,t.cssClass="ace-xcode",t.cssText=".ace-xcode .ace_gutter {background: #e8e8e8;color: #333}.ace-xcode .ace_print-margin {width: 1px;background: #e8e8e8}.ace-xcode {background-color: #FFFFFF;color: #000000}.ace-xcode .ace_cursor {color: #000000}.ace-xcode .ace_marker-layer .ace_selection {background: #B5D5FF}.ace-xcode.ace_multiselect .ace_selection.ace_start {box-shadow: 0 0 3px 0px #FFFFFF;border-radius: 2px}.ace-xcode .ace_marker-layer .ace_step {background: rgb(198, 219, 174)}.ace-xcode .ace_marker-layer .ace_bracket {margin: -1px 0 0 -1px;border: 1px solid #BFBFBF}.ace-xcode .ace_marker-layer .ace_active-line {background: rgba(0, 0, 0, 0.071)}.ace-xcode .ace_gutter-active-line {background-color: rgba(0, 0, 0, 0.071)}.ace-xcode .ace_marker-layer .ace_selected-word {border: 1px solid #B5D5FF}.ace-xcode .ace_keyword {font-weight:bold}.ace-xcode .ace_support.ace_function {font-weight:bold; color:#0000FF}.ace-xcode .ace_identifier {color: #408000}.ace-xcode .ace_constant.ace_language,.ace-xcode .ace_meta,.ace-xcode .ace_variable.ace_language {color: #6C05D3}.ace-xcode .ace_invisible {color: #BFBFBF}.ace-xcode .ace_constant.ace_character,.ace-xcode .ace_constant.ace_other {color: #275A5E}.ace-xcode .ace_constant.ace_numeric {color: #3A00DC}.ace-xcode .ace_entity.ace_other.ace_attribute-name,.ace-xcode .ace_support.ace_constant {color: #6C05D3}.ace-xcode .ace_fold {background-color: #C800A4;border-color: #000000}.ace-xcode .ace_entity.ace_name.ace_tag,.ace-xcode .ace_support.ace_class {color: #790EAD}.ace-xcode .ace_support.ace_type {color: #0000FF; font-style:italic}.ace-xcode .ace_storage {color: #C900A4}.ace-xcode .ace_string {}.ace-xcode .ace_comment {color: #4C4E4E}.ace-xcode .ace_indent-guide {background: url(data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAACCAYAAACZgbYnAAAAE0lEQVQImWP4////f4bLly//BwAmVgd1/w11/gAAAABJRU5ErkJggg==) right repeat-y}";var r=e("../lib/dom");r.importCssString(t.cssText,t.cssClass)})
```

**Comment by ylian** [comment]
If you have created an Ace theme file, send to me and I can include in the app.

**Comment by Tekl** [comment]
I've include the theme in my second post, but I'm not satisfied with the results. For testing purpose I've just overwritten theme-xcode.js

**Comment by ylian** [comment]
My bad. Didn't realize that was the Ace theme file already. I just gave it a try.

There are limitations to how much customization we can do with the Ace theme:
1. Ace doesn't support variable character width, so you will notice weird cursor positions around italic or bold fonts.
2. Syntax highlighting won't be exactly the same as the Script Editor. The Script Editor knows more information about the language thus can be more accurate with identifying tokens. Ace does an okay job but won't be as good.

I think the first point is a deal breaker in this case. But if you would be willing to sacrifice 'bold' and 'italics', I can still add a theme (basically just changing color schemes).

**Comment by Tekl** [comment]
1. The Github theme seems to work well with bold and italics.
2. Yes, so sad. ;-)

Just adding the colors would be fine. If I find time, I'll try to use Github as base.

**Comment by ylian** [comment]
The GitHub theme has issues with bold and italics as well. Move the cursor around bold or italic texts and you'll see the cursor position messed up quite often. But since the GitHub theme is acceptable, I guess it isn't a deal breaker after all.

I am adding the theme and will post here shortly.

**Comment by ylian** [comment]
I added a new theme "Script Editor". Colors are close. I think some details can be further improved. Some incorrect Ace types might be fixed by changing the `mode-applescript.js` file.

Let me know things that you would like to see improved. Thanks.

https://dl.dropboxusercontent.com/u/3445731/Quiver%202.1.3b9.zip

**Comment by ylian** [comment]
Made a few more enhancements. It will be never be perfect, but I think it's good enough for release.

https://dl.dropboxusercontent.com/u/3445731/Quiver%202.1.3b10.zip

**Comment by Tekl** [comment]
That's great. The problem with bold fonts can be fixed by using a font with a bold style like the more modern **Menlo**. The Monaco does not have a bold face so the letters gets bigger when computing faux bold.

Bzw. you should add the Script Editor theme in the correct alphabetical order.

Thanks for the great support and for your great app.

**Comment by ylian** [comment]
Wow, I didn't know that. Menlo looks much better. I have set it to the default font for the ace editor. Also added the Script Editor theme in the correct alphabetical order.

https://dl.dropboxusercontent.com/u/3445731/Quiver%202.1.3b11.zip

**Comment by Tekl** [comment]
That's great. Thanks.

**Comment by ylian** [comment]
This is implemented in Quiver 3.0, released today.
